# Check GPU status

Make surre to use : GPU runtime mode (Runtime->Change Runtime type -> python3 + GPU
)



In [1]:
# Check nvidia and nvcc cuda compiler

!nvidia-smi
!/usr/local/cuda/bin/nvcc --version

Sun Jun  7 09:27:10 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.82       Driver Version: 418.67       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   35C    P0    28W / 250W |      0MiB / 16280MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU  

#Mount Goolge Drive

In [2]:
# link to google drive

from google.colab import drive
drive.mount('/content/gdrive/')


Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/gdrive/


In [3]:
#check that Gdrive is mounted

!ls '/content/gdrive/My Drive/great learning/'

'capstone project'   model-1.11.h5
 glove.6B.100d.txt   model-1.17.h5
 glove.6B.200d.txt   model-1.32.h5
 glove.6B.300d.txt   model-1.56.h5
 glove.6B.50d.txt    model-2.06.h5
 glove.6B.zip	     model-2.22.h5
 images.npy	     News-Headlines-Dataset-For-Sarcasm-Detection
 model-0.93.h5	     SVHN_single_grey1.h5
 model-0.96.h5	     vgg_face
 model-1.06.h5


#Setup SSH port forwarding

In [4]:
#1 - setup ssh/user 


#Generate a random root password
import random, string
password = ''.join(random.choice(string.ascii_letters + string.digits) for i in range(30))


#Setup sshd
! apt-get install -qq -o=Dpkg::Use-Pty=0 openssh-server pwgen > /dev/null

#Set root password
! echo root:$password | chpasswd
! mkdir -p /var/run/sshd
! echo "PermitRootLogin yes" >> /etc/ssh/sshd_config
! echo "PasswordAuthentication yes" >> /etc/ssh/sshd_config
! echo “LD_LIBRARY_PATH=/usr/lib64-nvidia” >> /root/.bashrc
! echo “export LD_LIBRARY_PATH” >> /root/.bashrc

print("username: root")
print("password: ", password)

#Run sshd
get_ipython().system_raw('/usr/sbin/sshd -D &')


Creating config file /etc/ssh/sshd_config with new version
Creating SSH2 RSA key; this may take some time ...
2048 SHA256:pfdOVzvl4QAh2GjIXJwXL8wFMnrF2AVDVpL2ZBU6gOU root@e4f8a7aa92be (RSA)
Creating SSH2 ECDSA key; this may take some time ...
256 SHA256:1U/WnPddL3r3ff5M57wNf+Eqw78tC17V3CR//UIQz50 root@e4f8a7aa92be (ECDSA)
Creating SSH2 ED25519 key; this may take some time ...
256 SHA256:LOegy6EF99+AWeuYfnILo+nemlSsI4I+ghK/s/dgLsE root@e4f8a7aa92be (ED25519)
Created symlink /etc/systemd/system/sshd.service → /lib/systemd/system/ssh.service.
Created symlink /etc/systemd/system/multi-user.target.wants/ssh.service → /lib/systemd/system/ssh.service.
invoke-rc.d: could not determine current runlevel
invoke-rc.d: policy-rc.d denied execution of start.
username: root
password:  iJ3qr0Xq9sfXYTB3U75UxTiW5RTvcn


In [0]:
# 2 - Download Ngrok

! wget -q -c -nc https://bin.equinox.io/c/4VmDzA7iaHb/ngrok-stable-linux-amd64.zip
! unzip -qq -n ngrok-stable-linux-amd64.zip

In [6]:
# 3 - setup Ngrok - authtoken

#Ask token
print("Get your authtoken from https://dashboard.ngrok.com/auth")
import getpass
authtoken = getpass.getpass()

#Create tunnel
get_ipython().system_raw('./ngrok authtoken $authtoken && ./ngrok tcp 22 &')

Get your authtoken from https://dashboard.ngrok.com/auth
··········


In [9]:
print("Root password: {}".format(password))

Root password: iJ3qr0Xq9sfXYTB3U75UxTiW5RTvcn


In [11]:
! curl -s http://localhost:4040/api/tunnels | python3 -c \
"import sys, json; print(json.load(sys.stdin)['tunnels'][0]['public_url'])"

tcp://0.tcp.ngrok.io:19797




```
# This is formatted as code
```

You are ready to go. 
On Ngrok interface https://dashboard.ngrok.com/status you'll find the tcp address and the port 

connect using the following : 

```
ssh root@0.tcp.ngrok.io -p [ngrok_port]

> then enter the password generated previously

```

# Install SSH extension in VS Code.


# Open Command Palette (Ctrl + Shift + P) and Select Remote-SSH - Connect to Host
# Enter the address to the server. Example: root@0.tcp.ngrok.io:19797
# Enter the password generated above
# You will now be connected to the server

# Open Command terminal in VS Code (Ctrl + ~)
# type **export LD_PRELOAD=/usr/lib64-nvidia/libnvidia-ml.so** to link the GPU
# type **nvidia-smi** to check the available GPU
# Install tmux by using following command:
# **sudo apt-get install tmux**

# Before running any Python file just initiate the tmux session by typing tmux in command terminal
# Read more about tmux usage: https://www.hamvocke.com/blog/a-quick-and-easy-guide-to-tmux/

# You can access your drive by changing your work directory to /content/drive/

# You can also navigate via VS code folder explorer GUI 

# Using git commands or VS code git extension pull and push the git repos

# The colab sessions lasts only 12 hours. So save checkpoints to models.


In [0]:
# When done, kill Ngrok

!kill $(ps aux | grep './ngrok' | awk '{print $2}')